### ISSUE TITLE
Using locally installed rust toolchain

### ISSUE THREAD

**Issue by blackgnezdo** [bazelbuild/rules_rust#2275]
I am trying to use `rules_rust` on OpenBSD which is currently not supported. I read the docs to the point of discovering `rust_toolchain` which seems to fit the pattern. Sadly my attempt is currently failing with:
```
$ bazel build --extra_toolchains=//bazel:rust_openbsd :a
...
INFO: Build option --extra_toolchains has changed, discarding analysis cache.
Analyzing: target //:a (0 packages loaded, 0 targets configured)
ERROR: /home/greg/s/rop/bazel/BUILD:3:15: //bazel:rust_openbsd_impl: invalid label '/usr/local/bin/rustdoc' in attribute 'rust_doc' in 'rust_toolchain' rule: invalid target name '/usr/local/bin/rustdoc': target names may not start with '/'
ERROR: /home/greg/s/rop/bazel/BUILD:3:15: //bazel:rust_openbsd_impl: invalid label '/usr/local/bin/rust_std' in attribute 'rust_std' in 'rust_toolchain' rule: invalid target name '/usr/local/bin/rust_std': target names may not start with '/'
ERROR: /home/greg/s/rop/bazel/BUILD:3:15: //bazel:rust_openbsd_impl: invalid label '/usr/local/bin/rustc' in attribute 'rustc' in 'rust_toolchain' rule: invalid target name '/usr/local/bin/rustc': target names may not start with '/'
ERROR: /home/greg/s/rop/bazel/BUILD:3:15: //bazel:rust_openbsd_impl: invalid label '/usr/local/bin/rustc_lib' in attribute 'rustc_lib' in 'rust_toolchain' rule: invalid target name '/usr/local/bin/rustc_lib': target names may not start with '/'
ERROR: /home/greg/s/rop/bazel/BUILD:3:15: //bazel:rust_openbsd_impl: missing value for mandatory attribute 'rust_doc' in 'rust_toolchain' rule
ERROR: /home/greg/s/rop/bazel/BUILD:3:15: //bazel:rust_openbsd_impl: missing value for mandatory attribute 'rust_std' in 'rust_toolchain' rule
ERROR: /home/greg/s/rop/bazel/BUILD:3:15: //bazel:rust_openbsd_impl: missing value for mandatory attribute 'rustc' in 'rust_toolchain' rule
ERROR: /home/greg/s/rop/BUILD:3:12: While resolving toolchains for target //:a: invalid registered toolchain '//bazel:rust_openbsd': Error evaluating '//bazel:rust_openbsd': error loading package 'bazel': Package 'bazel' contains errors
ERROR: Analysis of target '//:a' failed; build aborted: 
```

The code is available at https://github.com/blackgnezdo/rop.

What's the easiest way to build rust programs with a local toolchain? 

If that's not the way to go, how do I extend `rules_rust` to add support for OpenBSD? I could conceivably host the bundle, or just keep a directory around and reference it with `local_repository`. What's the recipe for building the toolchain bundles?

**Comment by UebelAndre** [comment]
Something we could do is add a repository rule that allows you to set local paths to files that get symlinked into an external repository that can then be used in a toolchain. I'd be happy to review a PR for that 😄

**Comment by blackgnezdo** [comment]
OK, maybe I will feel inspired enough some day.

**Comment by solson** [comment]
I just ran into this issue as well — I'll copy what I said in #3509 for context:

> Is it possible to use the system Rust toolchain (i.e. from `PATH`) with `rules_rust`, akin to the auto-configured toolchains in `rules_cc` or `rules_java`, or like Buck2's `system_rust_toolchain`?
>
> I have a few reasons I want to support this (e.g. Rust provided from the outside by Nix, or in the context of a distro packaging build where downloading toolchain binaries is inappropriate or blocked by a sandbox), and I think my usecase is already covered for C++ and Java, but possibly not Rust.

----

> Something we could do is add a repository rule that allows you to set local paths to files that get symlinked into an external repository that can then be used in a toolchain. I'd be happy to review a PR for that 😄

@UebelAndre I'm potentially interested in contributing this, but I might not know enough about `rules_rust` and Bazel yet to get the design right. Would it also be accepted for `rules_rust` to contain an (optional) toolchain definition that resolves tool paths from `PATH`, akin to what `rules_cc` appears to provide?

The `rules_cc` implementation appears to be based on [`repository_ctx.which()`](https://bazel.build/rules/lib/builtins/repository_ctx#which):
- https://github.com/bazelbuild/rules_cc/blob/cbee84ad7f583049823f3d1497aab1264cf94f26/cc/private/toolchain/lib_cc_configure.bzl#L143-L154
- https://github.com/bazelbuild/rules_cc/blob/cbee84ad7f583049823f3d1497aab1264cf94f26/cc/private/toolchain/unix_cc_configure.bzl#L62-L87

And it looks like `rules_java` also uses `repository_ctx.which()` if `JAVA_HOME` isn't set:
- https://github.com/bazelbuild/rules_java/blob/02f488deec4eaf8188d8433762554dcc4e6083f7/toolchains/local_java_repository.bzl#L167

Rust doesn't have a `RUST_HOME` but there's a somewhat similar concept via looking up just the `rustc` binary and running `rustc --print=sysroot`, e.g.:

```console
$ ls "$(rustc --print=sysroot)/bin"
cargo  cargo-clippy  cargo-fmt  clippy-driver  rust-analyzer  rustc  rustdoc  rustfmt  rust-gdb  rust-gdbgui  rust-lldb

$ ls "$(rustc --print=sysroot)/lib/rustlib/src/rust/library/std/src"
alloc.rs  ascii.rs  backtrace  backtrace.rs  bstr.rs  collections  env.rs  error.rs  f128.rs  f16.rs  f32.rs  f64.rs  ffi  fs  fs.rs  hash  io  keyword_docs.rs  lib.miri.rs  lib.rs  macros.rs  net  num.rs  os  panicking.rs  panic.rs  path.rs  pat.rs  prelude  process  process.rs  random.rs  rt.rs  sync  sys  sys_common  test_helpers.rs  thread  time.rs
```
